workers.go

```diff
@@ -1,266 +1,221 @@
 	}
-	return iw.err
+	r.outChan = c
 }
 
 // SetDeadline allows a time to be set when the workers should stop.
 // Deadline needs to be handled by the IsDone method.
-func (iw *Worker) SetDeadline(t time.Time) *Worker {
-	iw.mu.Lock()
-	defer iw.mu.Unlock()
-	iw.Ctx, iw.cancel = context.WithDeadline(iw.Ctx, t)
-	return iw
+func (r *runner) SetDeadline(t time.Time) Runner {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.ctx, r.cancel = context.WithDeadline(r.ctx, t)
+	return r
 }
 
 // SetTimeout allows a time duration to be set when the workers should stop.
 // Timeout needs to be handled by the IsDone method.
-func (iw *Worker) SetTimeout(duration time.Duration) *Worker {
-	iw.mu.Lock()
-	defer iw.mu.Unlock()
-	iw.timeout = duration
-	return iw
-}
-
-// IsDone returns a context's cancellation or error
-func (iw *Worker) IsDone() <-chan struct{} {
-	return iw.Ctx.Done()
-}
-
-// SetWriterOut sets the writer for the Print* functions
-// (ex.
-// 		f, err := os.Create("output.txt"))
-//		defer f.Close()
-// 		worker.SetWriteOut(f)
-// )
-// If you have to print anything to stdout using the provided
-// Print functions can significantly improve performance by using
-// buffered output.
-func (iw *Worker) SetWriterOut(writer io.Writer) *Worker {
-	iw.writer.Reset(writer)
-	return iw
+func (r *runner) SetTimeout(duration time.Duration) Runner {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.timeout = duration
+	return r
+}
+
+func (r *runner) Wait() error {
+	r.requestStop()
+	r.wg.Wait()
+	if r.err != nil && !errors.Is(r.err, context.Canceled) {
+		return r.err
+	}
+	return nil
 }
 
-// Println prints line output of a
-func (iw *Worker) Println(a ...interface{}) {
-	iw.mu.Lock()
-	defer iw.mu.Unlock()
-	iw.internalBufferFlush()
-	_, _ = iw.writer.WriteString(fmt.Sprintln(a...))
+func (r *runner) StopRequested() <-chan struct{} {
+	return r.reqStopCtx.Done()
 }
 
-// Printf prints based on format provided and a
-func (iw *Worker) Printf(format string, a ...interface{}) {
-	iw.mu.Lock()
-	defer iw.mu.Unlock()
-	iw.internalBufferFlush()
-	_, _ = iw.writer.WriteString(fmt.Sprintf(format, a...))
+func (r *runner) IsDone() <-chan struct{} {
+	return r.ctx.Done()
 }
 
-// Print prints output of a
-func (iw *Worker) Print(a ...interface{}) {
-	iw.mu.Lock()
-	defer iw.mu.Unlock()
-	iw.internalBufferFlush()
-	_, _ = iw.writer.WriteString(fmt.Sprint(a...))
+// waitForSignal make sure we wait for a term signal and shutdown correctly
+func (r *runner) waitForSignal(signals ...os.Signal) {
+	go func() {
+		signal.Notify(r.signalChan, signals...)
+		<-r.signalChan
+		if r.cancel != nil {
+			r.cancel()
+		}
+	}()
 }
 
-// internalBufferFlush makes sure we haven't used up the available buffer
-// by flushing the buffer when we get below the danger zone.
-func (iw *Worker) internalBufferFlush() {
-	if iw.writer.Available() < internalBufferFlushLimit {
-		_ = iw.writer.Flush()
+func (r *runner) startWork() {
+	if r.err = r.beforeFunc(r.ctx); r.err != nil {
+		return
```

Comment:
So I removed the error chan - I think the request to stop does suffice what the errChan was doing, but I am not sure! 😅

Reply:
It's looking better.  The errchan may or may not be needed.  It was used as a blocker so the wait() would halt until it received an error wether it be legit or just a context.Canceled. Depending on how it all ends up it can just be ignored. I added a unit test to test finishing before the end that may help us figure out a rock solid solution.  I believe you are on the right track.